File: frontend/src/constants/montageTransitions.ts
export const MONTAGE_TRANSITION_OPTIONS = [
  { value: 'none', label: '硬切' },
  { value: 'fade', label: '淡入淡出' },
  { value: 'fadeblack', label: '黑场过渡' },
  { value: 'fadewhite', label: '闪白' },
  { value: 'wipeleft', label: '左划' },
  { value: 'wiperight', label: '右划' },
  { value: 'wipeup', label: '上划' },
  { value: 'wipedown', label: '下划' },
  { value: 'slideleft', label: '左滑' },
  { value: 'slideright', label: '右滑' },
  { value: 'slideup', label: '上滑' },
  { value: 'slidedown', label: '下滑' },
  { value: 'circleopen', label: '圆形展开' },
  { value: 'circleclose', label: '圆形收缩' },
  { value: 'dissolve', label: '溶解' },
] as const

export function estimateMontageDuration(
  segments: Array<{
    in_offset?: number
    out_offset?: number | null
    transition?: string
    transition_duration?: number
  }>,
  clipDurations: Map<string, number>,
  resolveKey: (segment: { clip_id: string; project_id?: string }) => string
): number {
  let total = 0
  segments.forEach((segment, index) => {
    const key = resolveKey(segment as { clip_id: string; project_id?: string })
    const clipDuration = clipDurations.get(key) || 0
    const inOffset = Number(segment.in_offset || 0)
    const outOffset =
      segment.out_offset === null || segment.out_offset === undefined
        ? clipDuration
        : Number(segment.out_offset)
    total += Math.max(0.1, outOffset - inOffset)
    if (index > 0 && segment.transition && segment.transition !== 'none') {
      total -= Number(segment.transition_duration ?? 0.5)
    }
  })
  return Math.max(0, total)
}

export function formatDuration(seconds: number): string {
  const s = Math.max(0, Math.round(seconds))
  const m = Math.floor(s / 60)
  const rest = s % 60
  return m > 0 ? `${m}分${rest}秒` : `${rest}秒`
}

File: backend/utils/montage_transitions.py
"""混剪转场常量"""

from typing import Dict, Optional, Set

# FFmpeg xfade 支持的转场名
XFADE_TRANSITIONS: Set[str] = {
    "fade",
    "fadeblack",
    "fadewhite",
    "wipeleft",
    "wiperight",
    "wipeup",
    "wipedown",
    "slideleft",
    "slideright",
    "slideup",
    "slidedown",
    "circleopen",
    "circleclose",
    "dissolve",
}

TRANSITION_LABELS: Dict[str, str] = {
    "none": "硬切",
    "fade": "淡入淡出",
    "fadeblack": "黑场过渡",
    "fadewhite": "闪白",
    "wipeleft": "左划",
    "wiperight": "右划",
    "wipeup": "上划",
    "wipedown": "下划",
    "slideleft": "左滑",
    "slideright": "右滑",
    "slideup": "上滑",
    "slidedown": "下滑",
    "circleopen": "圆形展开",
    "circleclose": "圆形收缩",
    "dissolve": "溶解",
}


def normalize_transition(value: Optional[str]) -> str:
    if not value or value == "none":
        return "none"
    return value if value in XFADE_TRANSITIONS else "fade"


def resolve_output_size(aspect_ratio: Optional[str]) -> tuple[int, int]:
    """返回 (width, height)。"""
    ratio = (aspect_ratio or "9:16").strip()
    if ratio == "16:9":
        return 1920, 1080
    return 1080, 1920
